libresin/libresin/core/shader.cpp

```diff
@@ -0,0 +1,160 @@
+#include <glad/gl.h>
+
+#include <chrono>
+#include <libresin/core/shader.hpp>
+#include <libresin/utils/exceptions.hpp>
+#include <libresin/utils/logger.hpp>
+#include <stdexcept>
+
+namespace resin {
+ShaderProgram::ShaderProgram(std::string_view name) : shader_name_(name), program_id_(glCreateProgram()) {
+  if (program_id_ == 0) {
+    throw std::runtime_error(std::format("Unable to create shader {}", name));
+  }
+}
+
+ShaderProgram::~ShaderProgram() { glDeleteProgram(program_id_); }
+
+void ShaderProgram::bind() const { glUseProgram(program_id_); }
+
+void ShaderProgram::unbind() const { glUseProgram(0); }  // NOLINT
+
+void ShaderProgram::recompile() {
+  using clock = std::chrono::high_resolution_clock;
+  auto start  = clock::now();
+
+  glDeleteProgram(program_id_);
+  program_id_ = glCreateProgram();
+  create_program();
+
+  auto stop     = clock::now();
+  auto duration = duration_cast<std::chrono::milliseconds>(stop - start);
+  Logger::debug("Shader {} recompilation took {}", shader_name_, duration);
+}
+
+void ShaderProgram::link_program() {
+  glLinkProgram(program_id_);
+
+  GLint status = 0;
+  glGetProgramiv(program_id_, GL_LINK_STATUS, &status);
```

Comment:
could be a static function that gets program id, `GL_LINK_STATUS`/`GL_VALIDATE_STATUS` and returns std::optional string or something, same for `glGetShaderiv`